### ISSUE TITLE
Collect use-cases requiring hierarchical parallelism

### ISSUE THREAD

**Issue by NaderAlAwar** [NVIDIA/cccl#8358]
## Description

We are planning on adding an API for hierarchical parallelism (#6410) to CUB and we would like your input to help identify use cases that would benefit from this API to help influence the design. These use cases should show instances where the existing CUB (or Thrust) API lacks certain abstractions for cooperative processing of a single logical element by multiple threads, synchronization between phases of work on related data, and composition of algorithmic stages with mismatched granularities such as group-level mask generation followed by element-level transformation.

Since these abstractions are missing, it should be possible to show that for the use cases we are looking for, a simple custom CUDA C++ kernel can outperform CUB currently.

## Use cases

Each use case listed should follow this pattern:

**Domain:** this can be either a library name, e.g., "cuDF", or a more general domain like "LLM Inference"
**Kernel name and description:** If the kernel/workload is well known and has a name, this can be listed here, e.g., rms norm. A short description of what the kernel does is also needed
**Proposed API:** pseudocode that describes how the algorithm will be called by the user
**Existing implementations:** If there is an existing implementation in a library, we want to compare it to an equivalent CUB implementation to see the performance difference between the two
**Performance measurements (optional):** If you can collect performance numbers comparing the CUB implementation to the custom CUDA C++ kernel that would help a lot

This issue can be closed once we have collected a sufficient number of uses cases that we can use to proceed with a PoC implementation of hierarchical parallelism

**Comment by NaderAlAwar** [comment]
## Segmented RMS norm

**Domain:** LLM Inference

**Kernel name and description:** The kernel applies root mean square (RMS) normalization over a contiguous 2D array, what we refer to in CUB as a segmented array. In the context of LLM inference, each segment is of size `hidden_size` and represents the hidden state at a given layer. For each segment, this kernel computes the RMS of the corresponding elements in a segment and then normalizes them by diving by the RMS. A more formal description is presented in the [pytorch docs](https://docs.pytorch.org/docs/stable/generated/torch.nn.modules.normalization.RMSNorm.html).

**Proposed API:**
```C++
    cub::DeviceHierarchicalTransform::Transform(
      d_input,  // Segmented array holding hidden state per segment
      d_output, 
      batch_size, // Number of segments, equivalent to the number of tokens
      hidden_size, // Size of each segment, fixed per model, e.g. 2880 for gpt-oss-120b
      rmsnorm_segment_op<T>{hidden_size, rms_norm_eps}, // Cooperative prolog op which calculates RMS per segment
      hierarchical_normalize_and_scale_op<T>{d_weight}); // Actual transform which takes output of prolog and normalizes
```

**Existing implementations:** The prominent implementations I've seen so far are:
* PyTorch: https://docs.pytorch.org/docs/stable/generated/torch.nn.modules.normalization.RMSNorm.html
* FlashInfer: https://docs.flashinfer.ai/generated/flashinfer.norm.rmsnorm.html#flashinfer.norm.rmsnorm. CUDA C++ implementation here https://github.com/flashinfer-ai/flashinfer/blob/e64ae8b8920034c2efc3f43326423ac5d8873b24/include/flashinfer/norm.cuh#L37.
  * There is also a [fused add](https://docs.flashinfer.ai/generated/flashinfer.norm.fused_add_rmsnorm.html) version which fuses the residual add that typically occurs prior to RMSNorm in transformers into the same kernel, as well as one with [fp8 quantization](https://docs.flashinfer.ai/generated/flashinfer.norm.fused_add_rmsnorm_quant.html)
  * There is also a [CuTe DSL implementation](https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/norm/kernels/rmsnorm.py). Interestingly, in the [PR it was added](https://github.com/flashinfer-ai/flashinfer/pull/2428), it is mentioned that compilation speed is an important reason for why it was added

> We prioritize using dsl for kernel development over cuda for faster JIT compilation speed. This PR is the first series that refactors the simple normalization kernels to cute-dsl. CUDA code should be ready to remove after we finish end-to-end testing.

* TensorRT-LLM: https://nvidia.github.io/TensorRT-LLM/_modules/tensorrt_llm/layers/normalization.html#RmsNorm. C++ implementations here https://github.com/NVIDIA/TensorRT-LLM/blob/fc2ba322d0e826e3f6716b292a6d199edb3cee83/cpp/tensorrt_llm/kernels/customAllReduceKernels.cu#L1993 (although these mostly seem to be the dused residual add variant)
* CuTile: https://github.com/NVIDIA/cutile-python/blob/c0eba4365b2ddbe3d1049150f2f24a1d27105834/test/bench_rms_norm.py#L43

## Performance measurements

<img width="2769" height="1257" alt="Image" src="https://github.com/user-attachments/assets/b1b5e14b-20fd-4432-8ef1-0bef6025e7bd" />

<img width="2769" height="1257" alt="Image" src="https://github.com/user-attachments/assets/753d9bd9-7f35-4e49-94e1-d03a422cedd6" />

This compares our implementation, C++ hierarchical POC final, to others. For both GPUs, we beat other implementations most of the time (`DeviceTransform` does not actually implement RMSNorm, it only loads and stores the same amount of data, and functions as an upper bound of what can be achieved). We are still worse for small hidden/segment sizes and the largest size on the B200. There are three main optimization opportunities left:

1. Use multiple segments per block for small segment sizes.
2. Use clusters for larger segment sizes instead of having one block reduce an entire segment.
3. Use pipelining or warp specialization. This would involve having one warp stage part of the segment, while other warps reduce it

The hierarchical parallelism implementation lives in my branch here https://github.com/NaderAlAwar/cccl/tree/hierarchical-parallelism commit 8b5f210124618f53993c5bf5115fad34dfe2c719. The benchmarks can be found under https://github.com/NaderAlAwar/cccl/tree/hierarchical-parallelism/cub/benchmarks/bench/hierarchical/rmsnorm. and the hierarchical parallelism benchmark showing our interface can be found here https://github.com/NaderAlAwar/cccl/blob/hierarchical-parallelism/cub/benchmarks/bench/hierarchical/rmsnorm/hierarchical_rmsnorm.cu

Note that I have not benchmarked the fused add variant, which is very widely used, but I expect similar performance results

**Comment by NaderAlAwar** [comment]
## cuDF bitmask kernels

**Domain:** Dataframe

**Kernel name and description:** cuDF stores column validity using packed null masks, where one bitmask_type word represents the validity of 32 rows. A recurring kernel pattern in cuDF computes one logical row result per thread, derives whether that row is valid, and then cooperatively materializes 32 row-validity decisions into one packed mask word. Many of these kernels also compute a valid-count side result, which is later converted into `null_count = num_rows - valid_count`.

This pattern appears in kernels such as:

  - `valid_if_kernel`: builds a packed validity mask from a unary predicate
  - `copy_if_else_kernel`: selects `lhs[i]` or `rhs[i]`, writes the selected value when valid, and builds the output null mask
  - `replace_kernel`: replaces matching values and builds output validity when nullability is involved
  - `concatenate_masks_kernel`: rebuilds the output validity mask for concatenated columns
  - `gpu_rolling`: computes rolling-window outputs and packs output validity
  - `get_json_object_kernel`: in its output pass, writes extracted JSON strings and builds output validity

In CUB terms, this is a hierarchical transform with a cooperative output epilogue: the transform produces one logical result per row, while the output side groups row-level validity into fixed-size segments of 32 and reduces those segment-local values into one bitmask_type word.

**Proposed API:**

```C++
cub::DeviceSegmentedTransform::TransformEpilog(
   cuda::std::make_tuple(lhs_iter, rhs_iter, filter_values),
   output_iterator,
   num_words,
   mask_word_bits,
   nullable_transform_op,
   epilog_op,
   stream);
```

Existing implementations: The prominent implementations in cuDF are custom kernels:

  - valid_if_kernel: cpp/include/cudf/detail/valid_if.cuh
  - copy_if_else_kernel: cpp/include/cudf/detail/copy_if_else.cuh
  - replace_kernel: cpp/src/replace/replace.cu
  - replace_nulls: cpp/src/replace/nulls.cu
  - concatenate_masks_kernel and fused_concatenate_kernel: cpp/src/copying/concatenate.cu
  - fused_concatenate_string_offset_kernel: cpp/src/strings/copying/concatenate.cu
  - gpu_rolling: cpp/src/rolling/detail/rolling.cuh
  - gpu_rolling_new: cpp/src/rolling/jit/kernel.cu
  - get_json_object_kernel: cpp/src/json/json_path.cu
  - copy_range_kernel: cpp/include/cudf/detail/copy_range.cuh
  - materialize_merged_bitmask_kernel: cpp/src/merge/merge.cu

**Comment by NaderAlAwar** [comment]
Following collection of performance data for RMSNorm, we are confident that the performance of the PoC is satisfactory.
